### PR TITLE
Bugfix/151 pmsmart

### PR DESCRIPTION
**Pull request by ckunki**
fixes issue https://github.com/itsallcode/white-rabbit/issues/151

### CODE REVIEW

File: launch/WhiteRabbit - JavaFXUI.launch [hunk 1]
```diff
@@ -7,7 +7,6 @@
         <listEntry value="1"/>
     </listAttribute>
     <listAttribute key="org.eclipse.debug.ui.favoriteGroups">
-        <listEntry value="org.eclipse.debug.ui.launchGroup.debug"/>
```
Comment (kaklakariada) [inline launch/WhiteRabbit - JavaFXUI.launch:10]: Please revert this change. This removes the debug launch config favorite

File: plugins/pmsmart/src/test/resources/log4j2-test.xml [hunk 2]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Configuration status="INFO">
+    <Appenders>
+        <Console name="Console" target="SYSTEM_OUT">
+            <PatternLayout charset="UTF-8" noConsoleNoAnsi="true"
+                pattern="%date{HH:mm:ss.SSS} [%thread] %highlight{%-5level} %logger{1.} - %message%n" />
+        </Console>
+        <RollingFile name="File" fileName="${sys:user.home}/.whiterabbit/logs/white-rabbit.log"
```
Comment (kaklakariada) [inline plugins/pmsmart/src/test/resources/log4j2-test.xml:8]: Do you log to the file by intention? Please remove the RollingFile appender if possible.

File: plugins/pmsmart/build.gradle [hunk 3]
```diff
@@ -8,6 +8,8 @@ dependencies {
     implementation "io.github.bonigarcia:webdrivermanager:4.4.3"
     
     testImplementation project(':logic')
+    testImplementation project(':logic').sourceSets.test.output
```
Comment (kaklakariada) [inline plugins/pmsmart/build.gradle:11]: This looks like a hack. If necessary, please move classes used by tests to a common test util project.

File: plugins/pmsmart/src/main/java/org/itsallcode/whiterabbit/plugin/pmsmart/web/page/DateSelector.java [hunk 4]
```diff
@@ -23,9 +30,15 @@ public DateSelector(Driver driver, Element selector)
 
     public void select(LocalDate day)
     {
+        if (day.isEqual(CRITICAL_DATE))
+        {
+            LOG.debug("Selecting critical date");
```
Comment (kaklakariada) [inline plugins/pmsmart/src/main/java/org/itsallcode/whiterabbit/plugin/pmsmart/web/page/DateSelector.java:35]: Please remove debug code

File: plugins/pmsmart/src/test/java/org/itsallcode/whiterabbit/plugin/pmsmart/integration/ConfigMock.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
+package org.itsallcode.whiterabbit.plugin.pmsmart.integration;
+
+import java.nio.file.Path;
+import java.time.Duration;
+import java.util.Locale;
+import java.util.Optional;
+
+import org.itsallcode.whiterabbit.logic.Config;
+
+public class ConfigMock implements Config
```
Comment (kaklakariada) [inline plugins/pmsmart/src/test/java/org/itsallcode/whiterabbit/plugin/pmsmart/integration/ConfigMock.java:10]: You can also use Mockito for creating mocks. Then you don't need a separate class.

File: plugins/pmsmart/src/test/java/org/itsallcode/whiterabbit/plugin/pmsmart/integration/JsonFileStorageMock.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.Optional;
+
+import javax.json.bind.Jsonb;
+
+import org.itsallcode.whiterabbit.api.model.MonthData;
+import org.itsallcode.whiterabbit.logic.storage.data.JsonFileStorage;
+import org.itsallcode.whiterabbit.logic.storage.data.JsonModelFactory;
+import org.itsallcode.whiterabbit.logic.storage.data.JsonbFactory;
+
+public class JsonFileStorageMock extends JsonFileStorage
```
Comment (kaklakariada) [inline plugins/pmsmart/src/test/java/org/itsallcode/whiterabbit/plugin/pmsmart/integration/JsonFileStorageMock.java:14]: Same here: use Mockito to create a mock.

File: plugins/pmsmart/src/test/java/org/itsallcode/whiterabbit/plugin/pmsmart/integration/Tester.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+import org.itsallcode.whiterabbit.plugin.pmsmart.PMSmartExporter;
+import org.itsallcode.whiterabbit.plugin.pmsmart.web.WebDriverFactory;
+
+/**
+ * Intended for manual tests
+ */
+public class Tester
+{
+    private static final String PROJECT_FILE = "c:/HOME/Tmp/White-Rabbit-Test-Data/projects.json";
+    private static final String TIME_RECORDING_FILE = "c:/HOME/Tmp/White-Rabbit-Test-Data/time-record.json";
```
Comment (kaklakariada) [inline plugins/pmsmart/src/test/java/org/itsallcode/whiterabbit/plugin/pmsmart/integration/Tester.java:37]: This depends on a local file. Either include it as a resource or define test data within the class.

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: plugins/pmsmart/src/test/resources/log4j2-test.xml (hunk 2 above)
Comment (ckunki) [reply plugins/pmsmart/src/test/resources/log4j2-test.xml]: done

File: launch/WhiteRabbit - JavaFXUI.launch (hunk 1 above)
Comment (ckunki) [reply launch/WhiteRabbit - JavaFXUI.launch]: done

File: plugins/pmsmart/build.gradle (hunk 3 above)
Comment (ckunki) [reply plugins/pmsmart/build.gradle]: done - seems to have become obsolete in the meantime anyway 😃

File: plugins/pmsmart/src/main/java/org/itsallcode/whiterabbit/plugin/pmsmart/web/page/DateSelector.java (hunk 4 above)
Comment (ckunki) [reply plugins/pmsmart/src/main/java/org/itsallcode/whiterabbit/plugin/pmsmart/web/page/DateSelector.java:35]: OK, done.
Is it OK to add sample code to class Tester?

File: plugins/pmsmart/src/test/java/org/itsallcode/whiterabbit/plugin/pmsmart/integration/ConfigMock.java (hunk 5 above)
Comment (ckunki) [reply plugins/pmsmart/src/test/java/org/itsallcode/whiterabbit/plugin/pmsmart/integration/ConfigMock.java]: Replaced by local static final class and narrow interfaces.

File: plugins/pmsmart/src/test/java/org/itsallcode/whiterabbit/plugin/pmsmart/integration/JsonFileStorageMock.java (hunk 6 above)
Comment (ckunki) [reply plugins/pmsmart/src/test/java/org/itsallcode/whiterabbit/plugin/pmsmart/integration/JsonFileStorageMock.java]: Replaced by local static final class and narrow interfaces.

File: plugins/pmsmart/src/test/java/org/itsallcode/whiterabbit/plugin/pmsmart/integration/Tester.java (hunk 7 above)
Comment (ckunki) [reply plugins/pmsmart/src/test/java/org/itsallcode/whiterabbit/plugin/pmsmart/integration/Tester.java:44]: I felt (and interpreted) the intention of this class to enable manual real world tests with actual user data.

A sample use case was based on issue #151:
* Strange behavior 
* hard to reproduce
* suspected sporadic failures 
 
When observing this I wanted to create a framework in order to test with real data as easy as possible. 
So a developer can easily feed in his real files and investigate the observed issue.

Creating all these data in the class seemed to be inefficient and future issues might require the same process again. 
Additional test cases did not seem to be not feasible either, since the behavior reported by issue #151 is an integration issue.

Would it be better to replace the paths by fake paths like "/path/to/your/time-recording/for/year-month.json"?

If this still conflicts with code policy then I'd rather prefer to remove file Tester.java and keep it outside the repo.
This is how I found it when starting investigation.

What do you suggest?

File: logic/src/main/java/org/itsallcode/whiterabbit/logic/service/project/ProjectFileProvider.java [hunk 8]
```diff
@@ -0,0 +1,8 @@
+package org.itsallcode.whiterabbit.logic.service.project;
+
+import java.nio.file.Path;
+
+public interface ProjectFileProvider
```
Comment (kaklakariada) [inline logic/src/main/java/org/itsallcode/whiterabbit/logic/service/project/ProjectFileProvider.java:5]: Is this really necessary? In my opinion it makes the code more complicated and harder to understand.

If your goal is to simplify tests (i.e. avoid implementing unnecessary mocks), I would prefer to use mockito and mock only the required methods of the `Config` interface.

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: plugins/pmsmart/src/main/java/org/itsallcode/whiterabbit/plugin/pmsmart/web/page/DateSelector.java (hunk 4 above)
Comment (kaklakariada) [reply plugins/pmsmart/src/main/java/org/itsallcode/whiterabbit/plugin/pmsmart/web/page/DateSelector.java]: Yes, because its useful for debugging. But the Tester should be self contained, i.e. have no dependencies to local files.

File: plugins/pmsmart/src/test/java/org/itsallcode/whiterabbit/plugin/pmsmart/integration/Tester.java (hunk 7 above)
Comment (kaklakariada) [reply plugins/pmsmart/src/test/java/org/itsallcode/whiterabbit/plugin/pmsmart/integration/Tester.java:62]: OK, now I understand your intention, and that's fine for me. Please add a comment to the class and use dummy paths to make the intention clear :)
Another option would be to read an `.gitignore`d custom property file pointing to the real file. But in this case this is too much effort.

File: logic/src/main/java/org/itsallcode/whiterabbit/logic/service/project/ProjectFileProvider.java (hunk 8 above)
Comment (ckunki) [reply logic/src/main/java/org/itsallcode/whiterabbit/logic/service/project/ProjectFileProvider.java:5]: Thanks - done.
Class comment had been provided already.
Great!

File: plugins/pmsmart/src/test/java/org/itsallcode/whiterabbit/plugin/pmsmart/integration/Tester.java (hunk 7 above)
Comment (ckunki) [reply plugins/pmsmart/src/test/java/org/itsallcode/whiterabbit/plugin/pmsmart/integration/Tester.java]: Thanks - done.
Class comment had been provided already.
Great!

File: logic/src/main/java/org/itsallcode/whiterabbit/logic/service/project/ProjectFileProvider.java (hunk 8 above)
Comment (ckunki) [reply logic/src/main/java/org/itsallcode/whiterabbit/logic/service/project/ProjectFileProvider.java:5]: Thanks for your question.
I thought a while about this before and now again.
And you're right: My approach was triggered by testing being hard to set up.

Is this really necessary? - no.
Does it make the code more complicated and harder to understand ?- yes, I agree. Sigh.

On the other hand I found it not very beautiful to pass a complex "config" to several services each being interested only in few members, hence creating more dependencies than necessary and using a unnecessarily large interface. 

Another option could be to let Config simply implement (or extend) interface ProjektFileProvider. 
Would you prefer this option?

Comment (kaklakariada) [reply logic/src/main/java/org/itsallcode/whiterabbit/logic/service/project/ProjectFileProvider.java:5]: Yet another option: the config values don't change at runtime. We could just pass them to the constructor and avoid the provider. That's even simpler ;)

Comment (ckunki) [reply logic/src/main/java/org/itsallcode/whiterabbit/logic/service/project/ProjectFileProvider.java]: Excellent idea, thanks!
Done.

File: logic/src/test/java/org/itsallcode/whiterabbit/logic/ConfigTest.java [hunk 9]
```diff
@@ -50,7 +50,7 @@ public Locale getLocale()
         @Override
         public Optional<Duration> getCurrentHoursPerDay()
         {
-            return null;
+            return Optional.ofNullable(null);
```
Comment (kaklakariada) [inline logic/src/test/java/org/itsallcode/whiterabbit/logic/ConfigTest.java:53]:
```suggestion
            return Optional.empty();
```

File: logic/src/test/java/org/itsallcode/whiterabbit/logic/service/AppServiceTest.java [hunk 10]
```diff
@@ -110,7 +110,7 @@ void setUp()
         appService = createAppService(appServiceCallback, workingTimeService);
         appService.setUpdateListener(updateListenerMock);
         modelFactory = new JsonModelFactory();
-        contractTerms = new ContractTermsService(configMock);
+        contractTerms = new ContractTermsService(TestingConfig.builder().build().getCurrentHoursPerDay());
```
Comment (kaklakariada) [inline logic/src/test/java/org/itsallcode/whiterabbit/logic/service/AppServiceTest.java:113]: Directly define hoursPerDay, e.g. `Duration.ofHours(8)`. Then you don't need to create the TestingConfig at all, and it's clear which value is used.

File: plugins/pmsmart/src/test/resources/log4j2-test.xml [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+        <Console name="Console" target="SYSTEM_OUT">
+            <PatternLayout charset="UTF-8" noConsoleNoAnsi="true"
+                pattern="%date{HH:mm:ss.SSS} [%thread] %highlight{%-5level} %logger{1.} - %message%n" />
+        </Console>
+    </Appenders>
+
+    <Loggers>
+        <Root level="debug">
+            <AppenderRef ref="Console" />
+            <AppenderRef ref="File" />
```
Comment (kaklakariada) [inline plugins/pmsmart/src/test/resources/log4j2-test.xml:13]: Please remove this reference to the non-existing appender :)
```suggestion
```

**Review by kaklakariada — CHANGES_REQUESTED** [review]
Only minor findings :)

File: logic/src/test/java/org/itsallcode/whiterabbit/logic/ConfigTest.java (hunk 9 above)
Comment (ckunki) [reply logic/src/test/java/org/itsallcode/whiterabbit/logic/ConfigTest.java]: done

File: logic/src/test/java/org/itsallcode/whiterabbit/logic/service/AppServiceTest.java (hunk 10 above)
Comment (ckunki) [reply logic/src/test/java/org/itsallcode/whiterabbit/logic/service/AppServiceTest.java]: Very good! 
Done.

File: plugins/pmsmart/src/test/resources/log4j2-test.xml (hunk 11 above)
Comment (ckunki) [reply plugins/pmsmart/src/test/resources/log4j2-test.xml]: Ups. thanks.
Done.

**Review by kaklakariada — APPROVED** [review]

**Comment by ckunki** [comment]
Yeeha!
